enhancements/cluster-logging/flow-control-api.md

```diff
@@ -0,0 +1,180 @@
+---
+title: flow-control-api
+authors:
+  - "@alanconway"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2021-02-05
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Flow control API for Logging
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+If the log collector cannot keep up with the rate that logs are written, some logging data will be lost.
+Presently the logging admin has little control over when or how logs get lost.
+
+This proposal defines an API to let cluster administrators to limit logging rates, or ignore some logs entirely.
+Logs will still be lost if the collector cannot keep up, but administrators have more control over *what* is lost.
+
+Rate limits mean that:
+
+* The cost and volume of logging can be predicted more accurately in advance.
+* Noisy containers cannot produce unbounded log traffic and unfairly drown out other containers.
+* High-value logs can be preferred over low-value logs by assigning higher rate limits.
+* Log streams that are not needed can be ignored, which reduces the load on the logging infrastructure.
+
+The API allows selected rates and policies to be applied by namespaces or pod labels.
+New types of selector that are added in future should also be applicable.
+
+## Motivation - FIXME
+
+This section is for explicitly listing the motivation, goals and non-goals of
+this proposal. Describe why the change is important and the benefits to users.
+
+### Goals - FIXME
+
+List the specific goals of the proposal. How will we know that this has succeeded?
+
+### Non-Goals - FIXME
+
+The following are not goals for this proposal but may to addressed by future proposals building on this one:
+
+- `block` policy: containers that exceed their rate limit are back-pressured by leaving their stdout/stderr unread until they are within the target rate.
+- combined rate limits: specify a total rate limit for a pipeline, enforced by dynamically adjusting rate limits on all containers contributing to the pipeline.
+
+## Proposal - FIXME
```

Comment:
Are these applied at the fluentd level?  Or at the conmon level?

Reply:
Probably a bit of both. The model I have in  my mind is that the collector is reading pipes or sockets from conmon rather than files from conmon and can push back by not reading. That means comon only needs to poll for write, and provide an option to use blocking pipes or sockets, it doesn't need complex config for controlling per-container rates & blocking - they can be imposed from outside.  I'm not sure that will be the way it ends up - there's thought needed.

Reply:
The doc is mainly intended to describe the API,  not the impl, but I agree at least some "possible impl" notes might help clarify

Reply:
I have been thinking that the only way to really do this well is in the kernel, which hopefully my presentation for container plumbers will be accepted where I can present that idea.